File: presto-base-jdbc/src/test/java/com/facebook/presto/plugin/jdbc/TestQueryBuilder.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.facebook.presto.plugin.jdbc;

import com.facebook.presto.spi.ColumnHandle;
import com.facebook.presto.spi.Domain;
import com.facebook.presto.spi.Range;
import com.facebook.presto.spi.SortedRangeSet;
import com.facebook.presto.spi.TupleDomain;
import com.facebook.presto.spi.type.Type;
import com.facebook.presto.spi.type.VarcharType;
import com.facebook.presto.sql.parser.SqlParser;
import com.facebook.presto.sql.tree.DefaultTraversalVisitor;
import com.facebook.presto.sql.tree.Query;
import com.facebook.presto.sql.tree.QuerySpecification;
import io.airlift.slice.Slice;
import org.testng.annotations.Test;

import java.nio.CharBuffer;
import java.util.HashMap;
import java.util.List;

import static com.facebook.presto.spi.TupleDomain.withColumnDomains;
import static com.facebook.presto.spi.type.BigintType.BIGINT;
import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
import static com.google.common.base.Charsets.UTF_8;
import static com.google.common.collect.ImmutableList.copyOf;
import static com.google.common.collect.ImmutableMap.of;
import static com.google.common.collect.Lists.newArrayList;
import static io.airlift.slice.Slices.utf8Slice;
import static org.testng.Assert.assertEquals;

public class TestQueryBuilder
{
    static final List<JdbcColumnHandle> COLUMN_HANDLES = copyOf(newArrayList(
            new JdbcColumnHandle("testConnector", "name", VARCHAR),
            new JdbcColumnHandle("testConnector", "loan-number", BIGINT),
            new JdbcColumnHandle("testConnector", "lease", BOOLEAN),
            new JdbcColumnHandle("testConnector", "rate", DOUBLE)
    ));

    @Test
    public void testQueryBuilderWhere()
    {
        TupleDomain<ColumnHandle> tupleDomain = withColumnDomains(of(
                COLUMN_HANDLES.get(0), Domain.singleValue(utf8Slice("Homer Simpson")),
                COLUMN_HANDLES.get(2), Domain.singleValue(false),
                COLUMN_HANDLES.get(3), Domain.create(SortedRangeSet.of(Range.greaterThan(10_000.0)), false)
        ));

        QueryBuilder queryBuilder = new QueryBuilder("\"");
        String sql = queryBuilder.buildSql("bank", "loan", "car-loan", COLUMN_HANDLES, tupleDomain);

        assertEquals(extractWhereClause(sql), "((\"lease\" = false) AND (\"rate\" > 10000.0))");
    }

    @Test
    public void testExtendedQueryBuilderWhere()
    {
        TupleDomain<ColumnHandle> tupleDomain = withColumnDomains(of(
                COLUMN_HANDLES.get(0), Domain.singleValue(utf8Slice("Marge Simpson")),
                COLUMN_HANDLES.get(2), Domain.singleValue(true),
                COLUMN_HANDLES.get(3), Domain.create(SortedRangeSet.of(Range.greaterThan(45_003.08)), false)
        ));

        QueryBuilder qbWithVarcharSupport = new QueryBuilder("\"")
        {
            @Override
            protected boolean supportPredicateOnType(Type type)
            {
                return super.supportPredicateOnType(type) || type.equals(VarcharType.VARCHAR);
            }

            @Override
            protected String encode(Type type, Object value)
            {
                if (type.equals(VarcharType.VARCHAR)) {
                    Slice slice = (Slice) value;
                    CharBuffer cb = UTF_8.decode(slice.toByteBuffer());
                    //Escape this string to prevent SQL injection.
                    return "'" + cb.toString() + "'";
                }

                return super.encode(type, value);
            }
        };
        String sql = qbWithVarcharSupport.buildSql("bank", "loan", "car-loan", COLUMN_HANDLES, tupleDomain);

        assertEquals(
                extractWhereClause(sql),
                "(((\"name\" = 'Marge Simpson') AND (\"lease\" = true)) AND (\"rate\" > 45003.08))");
    }

    static String extractWhereClause(String sql)
    {
        final String extractedWhereHolder[] = new String[1];

        //Use the Presto parser itself and extract the WHERE clause as a string.
        SqlParser sqlParser = new SqlParser();
        sqlParser.createStatement(sql).accept(new DefaultTraversalVisitor<Object, Object>()
        {
            @Override
            protected Object visitQuery(Query node, Object context)
            {
                extractedWhereHolder[0] = ((QuerySpecification) node.getQueryBody())
                        .getWhere().get().toString();
                return super.visitQuery(node, context);
            }
        }, new HashMap<>());

        return extractedWhereHolder[0];
    }
}

File: presto-victoriametrics/src/test/kotlin/com/victoriametrics/presto/TestQueryBuilder.kt
/**
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.victoriametrics.presto

import com.facebook.presto.spi.ColumnHandle
import com.facebook.presto.spi.predicate.Domain
import com.facebook.presto.spi.predicate.Range
import com.facebook.presto.spi.predicate.TupleDomain
import com.facebook.presto.spi.predicate.ValueSet
import com.facebook.presto.spi.type.TimestampType
import com.victoriametrics.presto.inject.TestVmModule
import com.victoriametrics.presto.model.VmColumnHandle
import okhttp3.HttpUrl
import org.assertj.core.api.Assertions.assertThat
import org.testng.annotations.Test

class TestQueryBuilder {
    val unit = TestVmModule.init().queryBuilder()

    /**
     * WHERE timestamp > TIMESTAMP '2019-10-26 21:40:10.646'
     */
    @Test
    fun testGreaterThan() {
        val range = Range.greaterThan(TimestampType.TIMESTAMP, 1572151210000L)
        val url = test(range)[0]

        assertThat(url.queryParameter("start")).isEqualTo("1572151210.001")
        assertThat(url.queryParameter("end")).isNull()
    }

    /**
     * WHERE timestamp >= TIMESTAMP '2019-10-26 21:40:10.646'
     */
    @Test
    fun testGreaterThanOrEqual() {
        val range = Range.greaterThanOrEqual(TimestampType.TIMESTAMP, 1572151210000L)
        val url = test(range)[0]

        assertThat(url.queryParameter("start")).isEqualTo("1572151210.000")
        assertThat(url.queryParameter("end")).isNull()
    }

    /**
     * WHERE timestamp < TIMESTAMP '2019-10-26 21:40:10.646'
     */
    @Test
    fun testLessThan() {
        val range = Range.lessThan(TimestampType.TIMESTAMP, 1572151210000L)
        val url = test(range)[0]

        assertThat(url.queryParameter("start")).isNull()
        assertThat(url.queryParameter("end")).isEqualTo("1572151209.999")
    }

    /**
     * WHERE timestamp <= TIMESTAMP '2019-10-26 21:40:10.646'
     */
    @Test
    fun testLessThanOrEqual() {
        val range = Range.lessThanOrEqual(TimestampType.TIMESTAMP, 1572151210000L)
        val url = test(range)[0]

        assertThat(url.queryParameter("start")).isNull()
        assertThat(url.queryParameter("end")).isEqualTo("1572151210.000")
    }

    /**
     * WHERE timestamp BETWEEN TIMESTAMP '2019-10-26 21:40:10.646' AND TIMESTAMP '2019-10-26 22:40:10.646';
     */
    @Test
    fun testBetween() {
        val range = Range.range(TimestampType.TIMESTAMP, 1572151210646L, true, 1572154810000L, true)
        val url = test(range)[0]

        assertThat(url.queryParameter("start")).isEqualTo("1572151210.646")
        assertThat(url.queryParameter("end")).isEqualTo("1572154810.000")
    }

    /**
     * WHERE timestamp NOT BETWEEN TIMESTAMP '2019-10-26 00:00:00.000' AND TIMESTAMP '2019-10-30 00:00:00.999';
     */
    @Test
    fun testNotBetween() {
        val range1 = Range.lessThan(TimestampType.TIMESTAMP, 1572151210000L)
        val range2 = Range.greaterThan(TimestampType.TIMESTAMP, 1572151210999L)
        val urls = test(range1, range2)

        assertThat(urls[0].queryParameter("start")).isNull()
        assertThat(urls[0].queryParameter("end")).isEqualTo("1572151209.999")
        assertThat(urls[1].queryParameter("start")).isEqualTo("1572151211.000")
        assertThat(urls[1].queryParameter("end")).isNull()
    }

    private fun test(range: Range, vararg ranges: Range): List<HttpUrl> {
        val timestampHandle: ColumnHandle = VmColumnHandle("timestamp")

        val values = ValueSet.ofRanges(range, *ranges)
        val domain: Domain = Domain.create(values, false)

        // val element = TupleDomain.ColumnDomain<ColumnHandle>(
        //     timestampHandle, domain
        // )
        // val listOf: List<TupleDomain.ColumnDomain<ColumnHandle>> = listOf(element)
        val map = mapOf(timestampHandle to domain)
        val constraint = TupleDomain.withColumnDomains(map)

        return unit.build(constraint)
    }

    @Test
    fun testPushdown() {
        // BuiltInFunctionHandle
        // CallExpression("BETWEEN", )
    }
}
